sequencer/src/external_event_handler.rs

```diff
@@ -0,0 +1,368 @@
+use std::{collections::BTreeSet, sync::Arc};
+
+use anyhow::{Context, Result};
+use async_compatibility_layer::channel::{Receiver, Sender};
+use async_std::task::{self, JoinHandle};
+use espresso_types::PubKey;
+use hotshot::types::BLSPubKey;
+use hotshot_types::traits::network::{BroadcastDelay, ConnectedNetwork};
+use serde::{Deserialize, Serialize};
+use url::Url;
+
+/// An external message that can be sent to or received from a node
+#[derive(Serialize, Deserialize, Clone)]
+pub enum ExternalMessage {
+    /// A request for a node to respond with its identifier
+    /// Contains the public key of the node that is requesting the roll call
+    RollCallRequest(BLSPubKey),
+
+    /// A response to a roll call request
+    /// Contains the identifier of the node
+    RollCallResponse(RollCallInfo),
+}
+
+/// Information about a node that is used in a roll call response
+#[derive(Serialize, Deserialize, Clone)]
+pub struct RollCallInfo {
+    // The public API URL of the node
+    pub public_api_url: Option<Url>,
+}
+
+/// The external event handler state
+pub struct ExternalEventHandler {
+    // The `RollCallInfo` of the node (used in the roll call response)
+    pub roll_call_info: RollCallInfo,
+
+    // The tasks that are running
+    pub tasks: Vec<JoinHandle<()>>,
+
+    // The outbound message queue
+    pub outbound_message_sender: Sender<OutboundMessage>,
+}
+
+// The different types of outbound messages (broadcast or direct)
+pub enum OutboundMessage {
+    Direct(Vec<u8>, PubKey),
+    Broadcast(Vec<u8>),
+}
+
+impl ExternalEventHandler {
+    /// Creates a new `ExternalEventHandler` with the given network and roll call info
+    pub fn new<N: ConnectedNetwork<PubKey>>(
+        network: Arc<N>,
+        roll_call_info: RollCallInfo,
+    ) -> Result<Self> {
+        // Create the outbound message queue
+        let (outbound_message_sender, outbound_message_receiver) =
+            async_compatibility_layer::channel::bounded(10);
```

Comment:
I think the event handling will be quick, but should the channel capacity be increased just to be on the safe side?